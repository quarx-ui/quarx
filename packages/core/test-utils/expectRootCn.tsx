import React, { FC } from 'react';
import { render } from '@testing-library/react';

export const expectRootCn = <T extends FC<{ className?: string, children: any }>>(Component: T, name: string) => {
    const className = 'test-class-name';
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    const { container: { firstChild: rootNode } } = render(<Component className={className} />);
    expect(rootNode).toHaveClass(className);
    expect(rootNode).toHaveClass(name);
};
