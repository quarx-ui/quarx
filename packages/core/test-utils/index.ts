export { expectRootCn } from './expectRootCn';
export * from './testBEM';
