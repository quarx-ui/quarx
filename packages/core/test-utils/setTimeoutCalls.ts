interface FakeSetTimeout {
    mock: {
        calls: Array<[() => void]>; // TODO Доработать типизацию
    };
}

/** При рендере jest использует setTimeout(_flushCallback, 0),
 * из-за чего проверка количества вызовов setTimeout происходит нестабильно.
 *
 * @return массив вызовов setTimeout без _flushCallback */
export const setTimeoutCalls = () => (setTimeout as unknown as FakeSetTimeout)
    .mock.calls
    .filter(([fn]) => !String(fn).includes('_flushCallback'));
