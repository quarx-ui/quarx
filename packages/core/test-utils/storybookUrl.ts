type PropsType = Record<string, string | boolean | number>;

interface UrlOptions {
    component: string;
    story?: string;
    props?: PropsType;
}

const reducePropsToURLString = (props?: PropsType) => {
    if (!props) {
        return '';
    }

    const keysOfProps = Object.keys(props);
    const valuesOfProps = Object.values(props);

    if (keysOfProps.length > 1) {
        return keysOfProps.reduce((prev, cur) => {
            const currentProp = `${cur}:${props[cur]}`;
            return `${prev};${currentProp}`;
        });
    }
    return `${keysOfProps[0]}:${valuesOfProps[0]}`;
};

export function storyUrl(options: UrlOptions) {
    const { component, story = 'sandbox', props } = options;
    const reducedProps = reducePropsToURLString(props);

    return `?path=/story/core-${component}--${story}&args=${reducedProps}`;
}
