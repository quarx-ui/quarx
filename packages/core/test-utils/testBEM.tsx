import React, { ComponentType } from 'react';
import { screen, render, cleanup } from '@testing-library/react';

export const expectPropsInClasses = (
    element: HTMLElement,
    qxClassname?: string,
) => (
    prop: string,
    value?: string | boolean,
) => {
    const qxSelector = qxClassname ?? element.classList[0];

    const valueIsBoolean = typeof value === 'boolean' || value === 'true' || value === 'false';
    const booleanClassname = value === true || value === 'true' ? `${qxSelector}_${prop}` : undefined;

    const expectedClassname = booleanClassname
        ?? (!valueIsBoolean
            ? `${qxSelector}_${prop}_${value}`
            : undefined);

    if (!expectedClassname) {
        return null;
    }

    return expect(element).toHaveClass(expectedClassname);
};

export const expectPropsMapInClasses = (element: HTMLElement, qxClassname?: string) => {
    const expectPropInClasses = expectPropsInClasses(element, qxClassname);

    return (expectedProps: Record<string, string | boolean>) => {
        Object.keys(expectedProps).forEach((prop) => {
            expectPropInClasses(prop, expectedProps[prop]);
        });
    };
};

export const testStyleParams = <
    S extends Record<string, any>,
    P extends Record<string, any> = Record<string, any>,
>(
    Component: ComponentType<P>,
    expectedDefaultParams: S,
    defaultProps: P = {} as P,
    qxClassname?: string,
) => (props: { [key in keyof S]: S[key][] }) => {
    describe('style params', () => {
        describe('default values', () => {
            const testId = 'testStyleParams-default';

            render(
                <Component
                    {...defaultProps}
                    data-testid={testId}
                />,
            );

            const element = screen.getByTestId(testId);

            Object.entries(expectedDefaultParams).forEach(([key, value]) => {
                it(`${key}`, () => {
                    expectPropsInClasses(element, qxClassname)(key, value);
                });
            });

            cleanup();
        });

        Object.entries(props).forEach(([propName, values]) => {
            describe(propName, () => {
                Object.values(values).forEach((value) => {
                    it(`${value}`, () => {
                        const testId = `testStyleParams-${propName}-${value}`;

                        const testProps = { [propName]: value };
                        const { asFragment } = render(
                            <Component
                                {...(defaultProps ?? {})}
                                {...testProps as P}
                                data-testid={testId}
                            />,
                        );

                        const element = screen.getByTestId(testId);
                        expectPropsMapInClasses(element, qxClassname)(testProps as S);
                        expect(asFragment()).toMatchSnapshot();
                    });
                });
            });
        });
    });
};
