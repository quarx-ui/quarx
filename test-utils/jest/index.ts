export { testRootCn } from './testRootCn';
export * from './testBEM';
