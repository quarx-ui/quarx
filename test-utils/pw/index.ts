export {
    compareStorySnapshots,
    compareVariantSnapshots,
    comparePropertySnapshots,
    comparePropertiesMapSnapshots,
} from './compareSnapshots';
