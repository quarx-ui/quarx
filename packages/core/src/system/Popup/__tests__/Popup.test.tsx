import { testStyleParams } from '@core/test-utils';
import {
    OmittedPopupStyleParams,
    Popup,
    POPUP_PAPER_REFERENCE,
    PopupProps,
    PopupStyleParams,
} from '@core';
import { cleanup, render } from '@testing-library/react';

class ResizeObserver {
    /* eslint-disable class-methods-use-this, @typescript-eslint/no-empty-function, lines-between-class-members */
    observe() {}
    unobserve() {}
    disconnect() {}
    /* eslint-enable class-methods-use-this, @typescript-eslint/no-empty-function, lines-between-class-members */
}

describe('Popup tests', () => {
    let root: null | HTMLDivElement = null;
    window.ResizeObserver = ResizeObserver;

    beforeEach(() => {
        cleanup();
        root = document.createElement('div');
        document.body.append(root);
    });

    testStyleParams<Omit<PopupStyleParams, keyof OmittedPopupStyleParams>, PopupProps>(
        Popup,
        {
            disableBackdrop: false,
            reference: POPUP_PAPER_REFERENCE.window,
        },
        {
            open: true,
            anchor: { current: root },
            onClickAway: () => undefined,
            children: 'Статус',
            disablePortal: true,
        },
    )({
        disableBackdrop: [false, true],
        reference: Object.values(POPUP_PAPER_REFERENCE),
    });

    const TestChildren = 'Popup content';
    it('default popup render', () => {
        const { getByText } = render((
            <Popup
                open
                anchor={{ x: 0, y: 0 }}
                onClickAway={() => undefined}
            >
                {TestChildren}
            </Popup>
        ));
        expect(getByText(TestChildren)).toBeInTheDocument();
    });

    it('hidden popup', () => {
        const { queryByText } = render((
            <Popup
                open={false}
                anchor={{ x: 0, y: 0 }}
                onClickAway={() => undefined}
            >
                {TestChildren}
            </Popup>
        ));
        expect(queryByText(TestChildren)).toBe(null);
    });

    it('disablePortal', () => {
        const { asFragment } = render((
            <div>
                <Popup
                    open
                    anchor={{ x: 0, y: 0 }}
                    onClickAway={() => undefined}
                    disablePortal
                >
                    {TestChildren}
                </Popup>
            </div>
        ));
        expect(asFragment()).toMatchSnapshot();
    });
});
