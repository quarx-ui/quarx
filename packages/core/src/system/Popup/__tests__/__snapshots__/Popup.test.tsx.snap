// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup tests disablePortal 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-hidden="false"
      class="QxBackdrop css-kd6k92-SxBackdrop-root"
      data-component="QxBackdrop"
      style="opacity: 1; transition: opacity 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    >
      <div
        class="QxPopup QxPopup_x_-8 QxPopup_y_-8 css-1n8lkmp-QxPopup-root"
      >
        Popup content
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Popup tests style params disableBackdrop false 1`] = `
<DocumentFragment>
  <div
    aria-hidden="false"
    class="QxBackdrop css-kd6k92-SxBackdrop-root"
    data-component="QxBackdrop"
    style="opacity: 1; transition: opacity 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  >
    <div
      class="QxPopup css-hgsm3k-QxPopup-root"
      data-testid="testStyleParams-disableBackdrop-false"
    >
      Статус
    </div>
  </div>
</DocumentFragment>
`;

exports[`Popup tests style params disableBackdrop true 1`] = `
<DocumentFragment>
  <div
    class="QxPopup QxPopup_disableBackdrop css-1sitpu4-QxPopup-root"
    data-testid="testStyleParams-disableBackdrop-true"
  >
    Статус
  </div>
</DocumentFragment>
`;
